### PR TITLE
Support `ProofRule::CONCAT_CPROP` in ALF

### PR DESCRIPTION
**Pull request by ajreynol**
Also updates the internal inference, which had been short circuiting the evaluation of the spliced constants.

### CODE REVIEW

File: src/theory/strings/theory_strings_utils.h [hunk 1]
```diff
@@ -70,6 +70,11 @@ Node mkPrefix(Node t, Node n);
  */
 Node mkSuffix(Node t, Node n);
 
+/**
+ * Returns (suf t n), which is (str.substr t (- (str.len t) n) (str.len t)).
```
Comment (HanielB) [inline src/theory/strings/theory_strings_utils.h:74]:
```suggestion
 * Returns (suf t n), which is (str.substr t (- (str.len t) n) n).
```

**Review by HanielB — APPROVED** [review]
LGTM, just a minor comment

File: proofs/alf/cvc5/programs/Strings.smt3 [hunk 2]
```diff
@@ -54,11 +54,16 @@
...
 )
 
-; Get the term corresponding to the suffix of term s of fixed length n.
+; Get the term corresponding to the suffix of term s starting from position n.
 (define skolem_suffix_rem ((U Type :implicit) (s (Seq U)) (n Int))
   (str.substr s n (- (str.len s) n))
 )
 
+; Get the term corresponding to the suffix of term s of length n.
+(define $str_suffix_len ((U Type :implicit) (s (Seq U)) (n Int))
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply proofs/alf/cvc5/programs/Strings.smt3:63]: Programs and definitions in the ALF signature are essentially the same category in my view, so I am updating the convention to use `$` for both.

[the remark this answers is not in the document]
Comment (HanielB) [reply proofs/alf/cvc5/programs/Strings.smt3:63]: Sounds good.

**Review by HanielB — APPROVED** [review]

**Pull request merged**